students/cfarris/lessons/lesson07/assignment/parallel.py

```diff
@@ -0,0 +1,243 @@
+                                      data_list,
+                                      start_time)),
+               threading.Thread(target=_read_csv,
+                                args=(db,
+                                      customers_db,
+                                      customers,
+                                      data_list,
+                                      start_time))]
+
+    for each_thread in threads:
+        each_thread.start()
+
+    for each_thread in threads:
+        each_thread.join()
+
+    print("data_list: ", data_list)
+    return data_list
+
+
+def _read_csv(db, collection, csv_file, data_list=None, start_time=0):
+    '''
+    Opens csv files, creates a list of dictionarys
+    with matching keys from header to values
+    returns the dictionary and saves it to different
+    mongodb collections based upon the
+    file name given to main.
+    '''
+
+    if data_list is None:
+        data_list = []
+
+    with open(csv_file, 'r', encoding='utf-8-sig') as csv_file:
+        csv_reader = csv.reader(csv_file, delimiter=',')
+        header = True
+        first_line = []
+        num_processed = 0
+        coll_of_dicts = []
+        count_before_running = collection.count()
+        for row in csv_reader:
+            if header:
+                header = False
+                first_line = [value.strip() for value in row]
+            else:
+                num_processed += 1
+                new_dict = {}
+                str_row = []
+                str_row = [value.strip() for value in row]
+                for i in range(len(str_row)):
+                    k = first_line[i]
+                    v = str_row[i]
+                    novelty_test = check_redundancy(db, k, v, collection)
+                    if novelty_test:
+                        new_dict[first_line[i]] = str_row[i]
+                coll_of_dicts.append(new_dict)
+        try:
+            collection.insert_many(coll_of_dicts)
+        except Exception as err:
+            print("error!:", err, collection.name, coll_of_dicts)
+        time_to_run = time.time() - start_time
+        results = (num_processed,
+                   count_before_running,
+                   collection.count(),
+                   time_to_run)
+        print(collection.name,
+              " finished: ",
+              num_processed,
+              count_before_running,
+              collection.count(),
+              time_to_run)
+        data_list.append(results)
+        return data_list
+
+
+def check_redundancy(db, unique_identifier_key,
+                     unique_identifier_value,
+                     collection):
+    """
+    For each row to be entered into a collection,
+    this checks to see if it is already there.
+    I cannot get this to work.
+    """
+
+    if collection.find({unique_identifier_key:
+                        unique_identifier_value}).count() > 0:
+        return False
+    return True
```

Comment:
The unique identifier is always unique, so you won't find duplicates. The way to check for redundancy with mongo is to assign the _id to your product or customer id as you enter it in the database. Then, mongo will complain if you try to put in the same one twice. If you let mongo assign the _id, mongo will assign a unique id to the second entry of the same thing, and not pay any attention that everything else is the same.